src/pool/pool_jemalloc.c

```diff
@@ -0,0 +1,397 @@
+    }
+
+    return ptr;
+}
+
+static umf_result_t je_free(void *pool, void *ptr) {
+    (void)pool; // unused
+    assert(pool);
+
+    if (ptr != NULL) {
+        dallocx(ptr, 0);
+    }
+
+    return UMF_RESULT_SUCCESS;
+}
+
+static void *je_calloc(void *pool, size_t num, size_t size) {
+    assert(pool);
+    size_t csize = num * size;
+    void *ptr = je_malloc(pool, csize);
+    if (ptr == NULL) {
+        // TLS_last_allocation_error is set by je_malloc()
+        return NULL;
+    }
+
+    memset(ptr, 0, csize); // TODO: device memory is not accessible by host
+    return ptr;
+}
+
+static void *je_realloc(void *pool, void *ptr, size_t size) {
+    assert(pool);
+    if (size == 0 && ptr != NULL) {
+        dallocx(ptr, 0);
+        TLS_last_allocation_error = UMF_RESULT_SUCCESS;
+        return NULL;
+    } else if (ptr == NULL) {
+        return je_malloc(pool, size);
+    }
+
+    jemalloc_memory_pool_t *je_pool = (jemalloc_memory_pool_t *)pool;
+    int flags = MALLOCX_ARENA(je_pool->arena_index);
+    void *new_ptr = rallocx(ptr, size, flags);
+    if (new_ptr == NULL) {
+        TLS_last_allocation_error = UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+        return NULL;
+    }
+
+    return new_ptr;
+}
+
+static void *je_aligned_alloc(void *pool, size_t size, size_t alignment) {
+    assert(pool);
+    jemalloc_memory_pool_t *je_pool = (jemalloc_memory_pool_t *)pool;
+    unsigned arena = je_pool->arena_index;
+    int flags = MALLOCX_ALIGN(alignment) | MALLOCX_ARENA(arena);
+    void *ptr = mallocx(size, flags);
+    if (ptr == NULL) {
+        TLS_last_allocation_error = UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+        return NULL;
+    }
+
+    return ptr;
+}
+
+static umf_result_t je_initialize(umf_memory_provider_handle_t provider,
+                                  void *params, void **out_pool) {
+    assert(provider);
+    assert(out_pool);
+    (void)params; // unused
+
+    extent_hooks_t *pHooks = &arena_extent_hooks;
+    size_t unsigned_size = sizeof(unsigned);
+    int err;
+
+    jemalloc_memory_pool_t *pool = malloc(sizeof(jemalloc_memory_pool_t));
+    if (!pool) {
+        return UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+    }
+
+    pool->provider = provider;
+
+    unsigned arena_index;
+    err =
+        mallctl("arenas.create", (void *)&arena_index, &unsigned_size, NULL, 0);
+    if (err) {
+        fprintf(stderr, "Could not create arena.\n");
+        goto err_free_pool;
+    }
+
+    // setup extent_hooks for newly created arena
+    char cmd[64];
+    snprintf(cmd, sizeof(cmd), "arena.%u.extent_hooks", arena_index);
+    err = mallctl(cmd, NULL, NULL, (void *)&pHooks, sizeof(void *));
+    if (err) {
+        snprintf(cmd, sizeof(cmd), "arena.%u.destroy", arena_index);
+        mallctl(cmd, NULL, 0, NULL, 0);
+        fprintf(stderr,
+                "Could not setup extent_hooks for newly created arena.\n");
+        goto err_free_pool;
+    }
+
+    pool->arena_index = arena_index;
+
+    Je_pool = pool;
```

Comment:
what if Je_pool is already initialized with something?

Reply:
@bratpiorka Done